exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/Serializer.java

```diff
@@ -0,0 +1,161 @@
+  public void serializeBool(ProtoFieldInfo field, boolean value) throws IOException {
+    if (!value) {
+      return;
+    }
+    writeBool(field, value);
+  }
+
+  protected abstract void writeBool(ProtoFieldInfo field, boolean value) throws IOException;
+
+  /** Serializes a protobuf {@code enum} field. */
+  public void serializeEnum(ProtoFieldInfo field, int enumNumber) throws IOException {
+    if (enumNumber == 0) {
+      return;
+    }
+    writeEnum(field, enumNumber);
+  }
+
+  protected abstract void writeEnum(ProtoFieldInfo field, int enumNumber) throws IOException;
+
+  /** Serializes a protobuf {@code uint32} field. */
+  public void serializeUInt32(ProtoFieldInfo field, int value) throws IOException {
+    if (value == 0) {
+      return;
+    }
+    writeUint32(field, value);
+  }
+
+  protected abstract void writeUint32(ProtoFieldInfo field, int value) throws IOException;
+
+  protected abstract void writeInt64(ProtoFieldInfo field, long value) throws IOException;
+
+  /** Serializes a protobuf {@code fixed64} field. */
+  public void serializeFixed64(ProtoFieldInfo field, long value) throws IOException {
+    if (value == 0) {
+      return;
+    }
+    writeFixed64(field, value);
+  }
+
+  protected abstract void writeFixed64(ProtoFieldInfo field, long value) throws IOException;
+
+  protected abstract void writeFixed64Value(long value) throws IOException;
+
+  /** Serializes a proto buf {@code double} field. */
+  public void serializeDouble(ProtoFieldInfo field, double value) throws IOException {
+    if (value == 0D) {
+      return;
+    }
+    writeDouble(field, value);
+  }
+
+  protected abstract void writeDouble(ProtoFieldInfo field, double value) throws IOException;
+
+  protected abstract void writeDoubleValue(double value) throws IOException;
+
+  /**
+   * Serializes a protobuf {@code string} field. {@code utf8Bytes} is the UTF8 encoded bytes of the
+   * string to serialize.
+   */
+  public void serializeString(ProtoFieldInfo field, byte[] utf8Bytes) throws IOException {
+    if (utf8Bytes.length == 0) {
+      return;
+    }
+    writeString(field, utf8Bytes);
+  }
+
+  protected abstract void writeString(ProtoFieldInfo field, byte[] utf8Bytes) throws IOException;
+
+  /** Serializes a protobuf {@code bytes} field. */
+  public void serializeBytes(ProtoFieldInfo field, byte[] value) throws IOException {
+    if (value.length == 0) {
+      return;
+    }
+    writeBytes(field, value);
+  }
+
+  protected abstract void writeBytes(ProtoFieldInfo field, byte[] value) throws IOException;
+
+  protected abstract void writeStartMessage(ProtoFieldInfo field, int protoMessageSize)
+      throws IOException;
+
+  protected abstract void writeEndMessage() throws IOException;
+
+  /** Serializes a protobuf embedded {@code message}. */
+  public void serializeMessage(ProtoFieldInfo field, Marshaler message) throws IOException {
+    writeStartMessage(field, message.getProtoSerializedSize());
+    message.writeTo(this);
```

Comment:
@jkwatson I liked the idea of wrapping an interface containing the `write` methods into a binary or json writer type. But I ran into a problem in that

- Writing of nested messages differs significantly between binary and json so it would need to be implemented in such a writer
- Writing a nested message requires calling back into the Serializer since it's basically just recursion. But a writer interface implementation is not able to call back into the concrete serializer if the serializer wraps it (well not without some circular references from parent to child)

I couldn't come up with a good solution that allows the Serializer to be concrete and extract extendible methods into a different interface. I could try again after implementing the JSON version where we might realize some of these extension points could be moved around. How does that sound?

Reply:
I think there is a way to do it, but we can definitely hold on refactoring until after the json version is implemented and things are a bit more clear.